pkgs/shells/hooks/sourceInputCompletionHook/source-input-completion-hook.sh

```diff
@@ -0,0 +1,40 @@
+
+# This function is meant to inspect the inputs to a derivation and source the shell completion.
+# This is intended to be only used with nix-shell to provide a better user experience, and is not
+# intended to be used within nixpkgs as shell completion is only useful in an interactive shell.
+sourceInputCompletion() {
+    local completionPath=
+
+    case $(basename $SHELL) in
+        bash)
+            completionPath=share/bash-completion/completions/
+            ;;
+        fish)
+            completionPath=share/fish/vendor_completions.d/
+            ;;
+        zsh)
+            completionPath=share/zsh/site-functions/
+            ;;
+        *)
+            echo "$(basename $SHELL) is not supported for shell completion. Skipping."
+            exit 0
+    esac
+
+    declare -A inputsSourced
+    for input in $nativeBuildInputs $buildInputs $propagatedNativeBuildInputs $propagatedBuildInputs; do
+        local completionDir=$input/$completionPath
+        if [ -d $completionDir ]; then
+            for script in $(@findutils@/bin/find $completionDir -type f); do
+                . $script
```

Comment:
Bash completions should be loaded lazily. This can be achieved with `XDG_DATA_DIRS` via #103501 or if that's too radical, a shell hook.

Reply:
Yea, I'm thinking of closing this in favor of #103501 . Made this PR before I was aware of it's existence